source/extensions/clusters/redis/redis_cluster.cc

```diff
@@ -0,0 +1,278 @@
+
+  active_query_ = parent_.dns_resolver_->resolve(
+      dns_address_, parent_.dns_lookup_family_,
+      [this](const std::list<Network::Address::InstanceConstSharedPtr>&& address_list) -> void {
+        active_query_ = nullptr;
+        ENVOY_LOG(trace, "async DNS resolution complete for {}", dns_address_);
+        parent_.redis_discovery_session_->registerDiscoveryAddress(address_list, port_);
+        parent_.redis_discovery_session_->startResolve();
+      });
+}
+
+// RedisCluster
+RedisCluster::RedisDiscoverySession::RedisDiscoverySession(
+    Envoy::Extensions::Clusters::Redis::RedisCluster& parent,
+    NetworkFilters::Common::Redis::Client::ClientFactory& client_factory)
+    : parent_(parent), dispatcher_(parent.dispatcher_),
+      resolve_timer_(parent.dispatcher_.createTimer([this]() -> void { startResolve(); })),
+      client_factory_(client_factory), buffer_timeout_(0) {}
+
+namespace {
+Network::Address::InstanceConstSharedPtr
+ProcessCluster(const NetworkFilters::Common::Redis::RespValue& value) {
+  ASSERT(value.type() == NetworkFilters::Common::Redis::RespType::Array);
+  ASSERT(value.asArray().size() >= 2);
+  std::string address = value.asArray()[0].asString();
+  bool ipv6 = (address.find(":") != std::string::npos);
+  if (ipv6) {
+    return std::make_shared<Network::Address::Ipv6Instance>(address,
+                                                            value.asArray()[1].asInteger());
+  }
+  return std::make_shared<Network::Address::Ipv4Instance>(address, value.asArray()[1].asInteger());
+}
+} // namespace
+
+RedisCluster::RedisDiscoverySession::~RedisDiscoverySession() {
+  if (current_request_) {
+    current_request_->cancel();
+    current_request_ = nullptr;
+  }
+
+  while (!client_map_.empty()) {
+    client_map_.begin()->second->client_->close();
+  }
+}
+
+void RedisCluster::RedisDiscoveryClient::onEvent(Network::ConnectionEvent event) {
+  if (event == Network::ConnectionEvent::RemoteClose ||
+      event == Network::ConnectionEvent::LocalClose) {
+    auto client_to_delete = parent_.client_map_.find(host_);
+    ASSERT(client_to_delete != parent_.client_map_.end());
+    parent_.dispatcher_.deferredDelete(std::move(client_to_delete->second->client_));
+    parent_.client_map_.erase(client_to_delete);
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::registerDiscoveryAddress(
+    const std::list<Envoy::Network::Address::InstanceConstSharedPtr>& address_list,
+    const uint32_t port) {
+  // Since the address from DNS does not have port, we need to make a new address that has port in
+  // it.
+  for (const Network::Address::InstanceConstSharedPtr& address : address_list) {
+    ASSERT(address != nullptr);
+    discovery_address_list_.push_back(Network::Utility::getAddressWithPort(*address, port));
+  }
+}
+
+void RedisCluster::RedisDiscoverySession::startResolve() {
+  parent_.info_->stats().update_attempt_.inc();
+  // If a resolution is currently in progress, skip it.
```

Comment:
When does this happen? Shouldn't we only start resolving when the last one completes?

Reply:
This can happened if the 2nd DNS entry got resolved while the startResolve is in-progress for the first DNS entry.